### ISSUE TITLE
Add a Warning Mechanism When Site Resources are Maximized

### ISSUE THREAD

**Issue by allanftd** [Automattic/wp-calypso#58041]
### What

Add a notification or warning (via email and dashboard) when site resources gets maximized, e.g. maximum database size, theme/plugin directory size, or disk space already reached

### Why

User on 32738155-hc had an issue with one page using WPDatatable plugin. The content does not show up when a visitor is not logged in when opening the page. After the user removed some tables (and reduced the size of some tables) in the plugin, the rest of the content is showing as normal. Speculating that the plugin reached maximum allowed directory size for plugins indicated in the Site Health page, it caused the issue. But it took a note from a visitor for them to know about this issue. Their feedback is if only they were sent a notice about this issue, they would have done proactive action to fix the issue. But they did not receive such notification at all.

### How

User is requesting us to add some warning mechanism to give users heads up in case certain site limits that would impact how their site works is reached.

**Comment by mreishus** [comment]
I'm not yet sure what specific limit this user ran into. Is it the overall disk quota or something else? I'm not seeing granular limits on the plugin or themes directory. The plugin, wpdatatables, does write to its own database tables, so maybe there is a DB size limit. Investigating that.
I also noticed that sometimes wpdatatables can cause a run-time out of memory error when trying to load large CSV tables.

**Comment by mreishus** [comment]
I'm finding difficulty determining what limit the user ran into. Reading the support chat, I see that visitors to the user's site saw a blank screen when retrieving a certain page/post that rendered a wpDataTable, and that the user was able to resolve the problem by removing some rows from the wpDataTable.

The closest I've come to recreating this is this example of a wpDataTable with ~40,000 rows:
`https://mmratomic3.wpco mstaging.com/2022/02/07/data2/` (remove space in url's domain name to access)
If we `wget` the url, it downloads 71M of html data, and the server does not error. However, loading it in the browser causes the browser to struggle with parsing and displaying so many rows.

I've not yet seen an example where the server does not return one specific page containing a wpDataTable, but if it did, I believe it would have to be an out-of-memory error while rendering the response.

The only other limits I've seen happen with wpDataTable is that it crashes with an out-of-memory error when attempting to import a CSV with > 100,000 rows.

To summarize:
* Static resource limits like disk space or database size - I do not believe these are being exceeded.
* Sending too much data for the browser to handle gracefully - This seems like more of an individual plugin being non-performant and not resources being exceeded. There's not a good way to measure this
* Out of memory errors - These happen during runtime, when code tries to allocate more memory than is available. So there's not a meter we can display that says 'You are using Xmb of Ymb allocated to your site', however, we could email a user whenever an out-of-memory error occurs. Maybe we already have some sort of alerting system like this?

**Comment by mreishus** [comment]
Resources limits I know exist:

 * `space_quota` / `space_used`
   * Enforcement: Users won't be able to exceed the platform disk quota, linux will stop writing to disk and manual intervention would be required.
   * Alerting: Can be checked manually in site health, but only updated once every 12 hours.
     * TODO: We should ensure that users know when they are running out of space
* out of memory
  * Enforcement: OOM Killer / Process dies
  * Alerting: Unsure if already exists. Should we add? Systems seems to keep an eye on it ( p9o2xV-1kz-p2 ), should we alert site owners as well?

Resource limits I am unsure about:

* Plugin and theme directory size
  * Do these have their own limit or are they just a part of the regular `space_quota` / `space_used` system?
* DB Limits
  *  Plugins can write to their own database tables. I assume this does not count against `space_used`. Does this have its own quota system?

@jmdodd, What do you see as the most important thing to fix here to help out both the customers and the Atomic team?

**Comment by mreishus** [comment]
- There is no plugin or theme specific quota, only the larger disk quota
- There is no DB quota yet, but one will be added in the future
  - Check how many sites are affected if we add the DB size used to their disk space used. How many sites have (disk + db) > 200GB (or whatever their quota is?)

**Comment by mreishus** [comment]
In #51594, we added blocking behavior for uploads that would bring an atomic site to be over the disk space limit.
In site health, uploads.php, and the calypso media guide, you can see how much disk space you have remaining.
In D75827-code, we did a check for sites using too much DB space.
I don't believe we have any other warnings to add at this time.
